pkg/datapath/linux/config/config.go

```diff
@@ -73,18 +73,28 @@ var (
 	}
 )
 
+// HeaderNodeDefinesMap is the type containing the key-value pairs representing
+// extra define directives for datapath node configuration.
+type HeaderNodeDefinesMap map[string]string
+
+// HeaderNodeDefinesFn is a function returning the key-value pairs representing
+// extra define directives for datapath node configuration.
+type HeaderNodeDefinesFn func() (HeaderNodeDefinesMap, error)
```

Comment:
Assuming this is a pure function, why do we need the `error` return here? I imagine that whatever provides the `header-node-defines` group will just error out if it can't create the correct headers?

Somewhat related, could `NodeExtraDefines` just be `[]HeaderNodeDefinesMap` instead?

Reply:
The idea of returning a function rather than the materialized defines here was to support values not known at provisioning time, and that could change over time. For instance, different configs are currently configured accessing some local node property through the node global variable (and the associated accessor functions), which has not yet been modularized, and is initialized during daemon initialization. Even if the local node object was modularized, though, we would still need to return a function so that we materialize the corresponding values only once the corresponding start hook has completed. Given that the function may rely on external values, it might fail, which is the reason why it also returns a possible error value.

Reply:
> Are there values that currently change over time?

Honestly I'm not sure about this. I saw that the config contains parameters such as IP addresses and the device MTU, which could theoretically change (I don't know though whether there's any logic to retrigger the generation, or an explicit agent restart is required for a change to take effect). That's why I wrote `[...] that could change over time`, but it was mostly a supposition. And I don't have plans to add any new config which would require that. 

I still think that the `values not known at provisioning time` part applies though, because we cannot retrieve and materialize parameters such as the Node IP from a provide function, but we need to defer that to a start hook (which is where the provided functions are executed with the current implementation). Unless we also rewrite these constants in the ELF, but that would likely only move the problem to a different place.

Reply:
node_config.h is rewritten when the host datapath gets regenerated (see current usage of `WriteNodeConfig()` through `Loader.Reinitialize()`). It is also called from `loader.hashDatapath()`, so we need to make sure the output of `WriteNodeConfig()` stays stable throughout multiple calls during agent init, otherwise things may break in surprising ways. Do the functions in `h.nodeExtraDefines` block until their data is available? Or should they error out in that case?

The goal is to eventually remove our runtime dependency on clang and plumbing in all values at load time using constant rewriting, no longer relying on writing out node_config.h at all.

Reply:
> Do the functions in h.nodeExtraDefines block until their data is available? Or should they error out in that case?

The idea of `h.nodeExtraDefines` is that it behaves similarly to the global header generation function, that is when it gets called all the dependencies have already been resolved (for instance by start hooks) and can be directly accessed (e.g., local node information).